src/coprocessor/codec/batch/lazy_column.rs

```diff
@@ -0,0 +1,552 @@
+        use cop_datatype::FieldTypeTp;
+
+        let mut column = LazyBatchColumn::raw_with_capacity(1000);
+
+        let mut datum_raw: Vec<u8> = Vec::new();
+        DatumEncoder::encode(&mut datum_raw, &[Datum::U64(0xDEADBEEF)], true).unwrap();
+
+        for _ in 0..1000 {
+            column.push_raw(datum_raw.as_slice());
+        }
+
+        let col_info = {
+            let mut col_info = tipb::schema::ColumnInfo::new();
+            col_info.as_mut_accessor().set_tp(FieldTypeTp::LongLong);
+            col_info
+        };
+        let tz = Tz::utc();
+
+        b.iter(|| {
+            let mut col = test::black_box(&column).clone();
+            col.decode(test::black_box(tz), test::black_box(&col_info))
+                .unwrap();
+            test::black_box(&col);
+        });
+    }
+
+    /// Bench performance of decoding a decoded lazy batch column.
+    ///
+    /// Note that there is a clone in the bench suite, whose cost should be excluded.
+    #[bench]
+    fn bench_lazy_batch_column_clone_and_decode_decoded(b: &mut test::Bencher) {
+        use crate::coprocessor::codec::datum::{Datum, DatumEncoder};
+        use cop_datatype::FieldTypeTp;
+
+        let mut column = LazyBatchColumn::raw_with_capacity(1000);
+
+        let mut datum_raw: Vec<u8> = Vec::new();
+        DatumEncoder::encode(&mut datum_raw, &[Datum::U64(0xDEADBEEF)], true).unwrap();
+
+        for _ in 0..1000 {
+            column.push_raw(datum_raw.as_slice());
+        }
+
+        let col_info = {
+            let mut col_info = tipb::schema::ColumnInfo::new();
+            col_info.as_mut_accessor().set_tp(FieldTypeTp::LongLong);
+            col_info
+        };
+        let tz = Tz::utc();
+
+        column.decode(tz, &col_info).unwrap();
+
+        b.iter(|| {
+            let mut col = test::black_box(&column).clone();
+            col.decode(test::black_box(tz), test::black_box(&col_info))
+                .unwrap();
+            test::black_box(&col);
+        });
+    }
+
+    /// A vector based LazyBatchColumn
+    #[derive(Clone)]
+    enum VectorLazyBatchColumn {
+        Raw(Vec<Vec<u8>>),
+        Decoded(VectorValue),
+    }
+
+    impl VectorLazyBatchColumn {
+        #[inline]
+        pub fn raw_with_capacity(capacity: usize) -> Self {
+            VectorLazyBatchColumn::Raw(Vec::with_capacity(capacity))
+        }
+
+        #[inline]
+        pub fn clear(&mut self) {
+            match self {
+                VectorLazyBatchColumn::Raw(ref mut v) => v.clear(),
+                VectorLazyBatchColumn::Decoded(ref mut v) => v.clear(),
+            }
+        }
+
+        #[inline]
+        pub fn push_raw(&mut self, raw_datum: &[u8]) {
+            match self {
+                VectorLazyBatchColumn::Raw(ref mut v) => v.push(raw_datum.to_vec()),
+                VectorLazyBatchColumn::Decoded(_) => panic!(),
+            }
+        }
+    }
+
+    /// Bench performance of push 10 bytes to a vector based LazyBatchColumn.
```

Comment:
pushing

Reply:
Thanks a lot. In order to avoid a lot of rebase operations, I modified it in #4215 : https://github.com/tikv/tikv/pull/4215/commits/58ac02ffe755c1647027480edcec9ddfc52c5754  So when #4215 instead of this PR got merged your comments will be addressed.

Reply:
Got it.